### PR TITLE
feat(techdocs): add edit button support for bitbucketServer

### PR DESCRIPTION
**Pull request by omerfarukdogan**
Signed-off-by: Ömer Faruk Doğan <8117265+omerfarukdogan@users.noreply.github.com>

## Hey, I just made a Pull Request!

Fixed `BitbucketServerIntegration:resolveEditUrl` and enabled edit button support.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by camilaibs — APPROVED** [review]
The changes to the TechDocs files look great, thanks for contributing

**Review by benjdlambert — APPROVED** [review]
Nice work, integration changes looks good. Maybe we can get a @pjungermann signoff here too as codeowner for some of the new bitbucket stuff

File: plugins/techdocs-node/src/stages/generate/helpers.ts
Comment (pjungermann) [inline plugins/techdocs-node/src/stages/generate/helpers.ts:108]: Bitbucket Cloud supports edit URLs, too, but out of scope here.

Maybe I pick that up for Bitbucket Cloud.

File: plugins/techdocs-module-addons-contrib/src/ReportIssue/ReportIssue.tsx
Comment (pjungermann) [inline plugins/techdocs-module-addons-contrib/src/ReportIssue/ReportIssue.tsx:124]: Instead of explicitly including all the others (allow list), maybe it better to invert this (block list) so it only needs to be adjusted if some integration does not support it?

And what is the reason for not supporting the ReportIssueAddon?

**Comment by pjungermann** [comment]
changes overall look good 👍 

just a clarifying question

File: plugins/techdocs-module-addons-contrib/src/ReportIssue/ReportIssue.tsx
Comment (camilaibs) [reply plugins/techdocs-module-addons-contrib/src/ReportIssue/ReportIssue.tsx:124]: > And what is the reason for not supporting the ReportIssueAddon?
If I'm not wrong, the `IssueLink` component doesn't know how to create the URL for bitbuckt here:
https://github.com/backstage/backstage/blob/ad35364e971bf21d26e7b8ec3d74a00323996268/plugins/techdocs-module-addons-contrib/src/ReportIssue/IssueLink.tsx#L66-L69

Comment (omerfarukdogan) [reply plugins/techdocs-module-addons-contrib/src/ReportIssue/ReportIssue.tsx:124]: Bitbucket Server doesn't have an integrated Issue Tracker system. Therefore `ReportIssueAddon` is not applicable here.

Reading the method below, I assumed that `ReportIssueAddon` was developed to support only Github and Gitlab. That's why I added the condition that way.

https://github.com/backstage/backstage/blob/ad35364e971bf21d26e7b8ec3d74a00323996268/plugins/techdocs-module-addons-contrib/src/ReportIssue/IssueLink.tsx#L50-L55

Comment (pjungermann) [reply plugins/techdocs-module-addons-contrib/src/ReportIssue/ReportIssue.tsx:124]: I see. Thanks for the clarification. Bitbucket Cloud would have an integrated issue tracker, but I guess many use Jira instead though which got more integrated there as well some time ago.

The allow list makes perfect sense then!

**Review by pjungermann — APPROVED** [review]

**Comment by benjdlambert** [comment]
Thanks for this! :pray:

**Comment by pjungermann** [comment]
@omerfarukdogan not sure if Bitbucket Server has the same, but Bitbucket Cloud supports deep links to the online editor via `?mode=edit` which is different to how GitHub etc. do it (path segment with either blob, tree or edit).

In order to make it work for TechDocs, you have removed all query params so that the base URL appears and the relative path to the current page can be appended (e.g., `[...]/docs/` -> `[...]/docs/index.md`).

This `resolveEditUrl` functionality is also used at the entity page though. There it will not open the edit mode anymore but just a deep link to the view mode.

Unfortunately, mkdocs is not supporting such query params (only to append the path to those or a hash, not to insert the path in front of it).

Maybe an alternative could be to strip of the query string at the helper.ts at the techdocs module instead of doing it at the integration level. 🤔

**Comment by pjungermann** [comment]
see also https://github.com/mkdocs/mkdocs/issues/1378

**Pull request merged**
